cudax/include/cuda/experimental/__utility/driver_api.cuh

```diff
@@ -0,0 +1,84 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__UTILITY_DRIVER_API
+#define _CUDAX__UTILITY_DRIVER_API
+
+#include <cuda.h>
+
+#include <cuda/std/__exception/cuda_error.h>
+
+// Get the driver function by name using this macro
+#define CUDAX_GET_DRIVER_FUNCTION(function_name) \
+  reinterpret_cast<decltype(function_name)*>(get_driver_entry_point(#function_name))
+
+namespace cuda::experimental::detail::driver
+{
+inline void* get_driver_entry_point(const char* name)
+{
+  void* fn;
+  cudaDriverEntryPointQueryResult result;
+  cudaGetDriverEntryPoint(name, &fn, cudaEnableDefault, &result);
+  if (result != cudaDriverEntryPointSuccess)
+  {
+    if (result == cudaDriverEntryPointVersionNotSufficent)
+    {
+      ::cuda::__throw_cuda_error(cudaErrorNotSupported, "Driver does not support this API");
+    }
+    else
+    {
+      ::cuda::__throw_cuda_error(cudaErrorUnknown, "Failed to access driver API");
+    }
+  }
+  return fn;
+}
+
+template <typename Fn, typename... Args>
+inline void call_driver_fn(Fn fn, const char* err_msg, Args... args)
+{
+  CUresult status = fn(args...);
+  if (status != CUDA_SUCCESS)
+  {
+    ::cuda::__throw_cuda_error(static_cast<cudaError_t>(status), err_msg);
+  }
+}
+
+inline void ctxPush(CUcontext ctx)
+{
+  static auto driver_fn = CUDAX_GET_DRIVER_FUNCTION(cuCtxPushCurrent);
+  call_driver_fn(driver_fn, "Failed to push context", ctx);
+}
```

Comment:
why are we dynamically loading these functions instead of including `<cuda.h>` and linking to `libcuda`?

Reply:
We would need to require -lcuda compilation flag otherwise. This is more in line with the current CUDA runtime which does not require the compilation flag. There are compatibility reasons why current CUDA runtime does that and we probably want the same thing

Reply:
directly linking to `libcuda.so` means that any consuming library would only run on machines with the CUDA driver installed. This would mean that any application with runtime logic to dispatch to CUDA vs CPU based on HW support would fail to **load** when launched on a machine without the CUDA driver.

From a build engineer standpoint linking to libcuda.so should never happen

Reply:
cool thanks. i knew there must be a reason. TIL